File: speedviewlib/src/main/java/com/github/anastr/speedviewlib/LinearGauge.java
package com.github.anastr.speedviewlib;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.util.AttributeSet;

/**
 * this Library build By Anas Altair
 * see it on <a href="https://github.com/anastr/SpeedView">GitHub</a>
 */
public abstract class LinearGauge extends Gauge {

    private Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
    /** to draw part of bitmap */
    private Rect rect = new Rect();
    private Bitmap foregroundBitmap;

    /** horizontal or vertical direction */
    private Orientation orientation = Orientation.HORIZONTAL;

    public LinearGauge(Context context) {
        this(context, null);
    }

    public LinearGauge(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public LinearGauge(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initAttributeSet(context, attrs);
    }

    /**
     * update background and foreground bitmap,
     * this method called when change size, color, orientation...
     * <p>
     *     must call {@link #createBackgroundBitmapCanvas()} and
     *     {@link #createForegroundBitmapCanvas()} inside this method.
     * </p>
     */
    protected abstract void updateFrontAndBackBitmaps();

    private void initAttributeSet(Context context, AttributeSet attrs) {
        if (attrs == null)
            return;
        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.LinearGauge, 0, 0);

        int orientation = a.getInt(R.styleable.LinearGauge_sv_orientation, -1);
        if (orientation != -1)
            setOrientation(Orientation.values()[orientation]);
        a.recycle();
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldW, int oldH) {
        super.onSizeChanged(w, h, oldW, oldH);

        updateBackgroundBitmap();
    }

    @Override
    protected void updateBackgroundBitmap() {
        updateFrontAndBackBitmaps();
    }

    protected final Canvas createForegroundBitmapCanvas() {
        if (getWidthPa() == 0 || getHeightPa() == 0)
            return new Canvas();
        foregroundBitmap = Bitmap.createBitmap(getWidthPa(), getHeightPa(), Bitmap.Config.ARGB_8888);
        return new Canvas(foregroundBitmap);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        if (orientation == Orientation.HORIZONTAL)
            rect.set(0, 0
                    , (int)(getWidthPa() * getOffsetSpeed()), getHeightPa());
        else
            rect.set(0, getHeightPa() - (int)(getHeightPa() * getOffsetSpeed())
                    , getWidthPa(), getHeightPa());

        canvas.translate(getPadding(), getPadding());
        canvas.drawBitmap(foregroundBitmap, rect, rect, paint);
        canvas.translate(-getPadding(), -getPadding());

        drawSpeedUnitText(canvas);
    }

    public Orientation getOrientation() {
        return orientation;
    }

    /**
     * change fill orientation,
     * this will change view width and height.
     * @param orientation new orientation.
     */
    public void setOrientation(Orientation orientation) {
        this.orientation = orientation;
        if (!isAttachedToWindow())
            return;
        requestLayout();
        updateBackgroundBitmap();
        invalidate();
    }

    public enum Orientation {
        HORIZONTAL, VERTICAL
    }
}

File: speedviewlib/src/main/java/com/github/anastr/speedviewlib/LinearGauge.kt
package com.github.anastr.speedviewlib

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.util.AttributeSet

/**
 * this Library build By Anas Altair
 * see it on [GitHub](https://github.com/anastr/SpeedView)
 */
abstract class LinearGauge @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : Gauge(context, attrs, defStyleAttr) {

    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    /** to draw part of bitmap  */
    private val rect = Rect()
    private var foregroundBitmap: Bitmap = Bitmap.createBitmap(1, 1, Bitmap.Config.ARGB_8888)

    /**
     * horizontal or vertical direction .
     * change fill orientation,
     * this will change view width and height.
     */
    var orientation = Orientation.HORIZONTAL
        set(orientation) {
            field = orientation
            if (isAttachedToWindow) {
                requestLayout()
                invalidateGauge()
            }
        }

    init {
        initAttributeSet(context, attrs)
    }

    /**
     * update background and foreground bitmap,
     * this method called when change size, color, orientation...
     *
     *
     * must call [createBackgroundBitmapCanvas] and
     * [createForegroundBitmapCanvas] inside this method.
     *
     */
    protected abstract fun updateFrontAndBackBitmaps()

    private fun initAttributeSet(context: Context, attrs: AttributeSet?) {
        if (attrs == null)
            return
        val a = context.theme.obtainStyledAttributes(attrs, R.styleable.LinearGauge, 0, 0)

        val orientation = a.getInt(R.styleable.LinearGauge_sv_orientation, -1)
        if (orientation != -1)
            this.orientation = Orientation.values()[orientation]
        a.recycle()
    }

    override fun onSizeChanged(w: Int, h: Int, oldW: Int, oldH: Int) {
        super.onSizeChanged(w, h, oldW, oldH)

        updateBackgroundBitmap()
    }

    override fun updateBackgroundBitmap() {
        updateFrontAndBackBitmaps()
    }

    protected fun createForegroundBitmapCanvas(): Canvas {
        if (widthPa == 0 || heightPa == 0)
            return Canvas()
        foregroundBitmap = Bitmap.createBitmap(widthPa, heightPa, Bitmap.Config.ARGB_8888)
        return Canvas(foregroundBitmap)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        if (this.orientation == Orientation.HORIZONTAL)
            rect.set(0, 0, (widthPa * getOffsetSpeed()).toInt(), heightPa)
        else
            rect.set(0, heightPa - (heightPa * getOffsetSpeed()).toInt(), widthPa, heightPa)

        canvas.translate(padding.toFloat(), padding.toFloat())
        canvas.drawBitmap(foregroundBitmap, rect, rect, paint)
        canvas.translate((-padding).toFloat(), (-padding).toFloat())

        drawSpeedUnitText(canvas)
    }

    enum class Orientation {
        HORIZONTAL, VERTICAL
    }
}
